AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine.swift

```diff
@@ -238,115 +238,106 @@ final class StorageEngine: StorageEngineBehavior {
+                if self.syncEngine == nil {
+                    log.error("Unable to sync because syncEngine is nil")
+                }
+                completion(.success(deletedModel))
+                return
             }
-            if self.syncEngine == nil {
-                log.error("Unable to sync because syncEngine is nil")
+
+            guard #available(iOS 13.0, *) else {
+                completion(.success(deletedModel))
+                return
             }
-            completion(.success(deletedModel))
-            return
-        }
 
-        if #available(iOS 13.0, *) {
-            let syncCompletionWrapper: DataStoreCallback<Void> = { syncResult in
-                switch syncResult {
+            syncDeletions(of: modelType,
+                          modelSchema: modelSchema,
+                          withModels: [deletedModel],
+                          associatedModels: queryAndDeleteResult.associatedModelsMap,
+                          syncEngine: syncEngine) {
+                switch $0 {
                 case .success:
                     completion(.success(deletedModel))
                 case .failure(let error):
                     completion(.failure(error))
                 }
             }
-            self.syncDeletions(of: modelType,
-                               modelSchema: modelSchema,
-                               withModels: [deletedModel],
-                               associatedModels: associatedModelsFromTransactionResult,
-                               syncEngine: syncEngine,
-                               completion: syncCompletionWrapper)
-        } else {
-            completion(.success(deletedModel))
+        case .failure(let error):
+            completion(.failure(error))
         }
     }
 
     func delete<M: Model>(_ modelType: M.Type,
                           modelSchema: ModelSchema,
-                          predicate: QueryPredicate,
+                          filter: QueryPredicate,
                           completion: @escaping DataStoreCallback<[M]>) {
         let transactionResult = queryAndDeleteTransaction(modelType,
                                                           modelSchema: modelSchema,
-                                                          deleteInput: .withPredicate(predicate: predicate))
-        let modelsFromTransactionResult = collapseMResult(transactionResult)
-        let associatedModelsFromTransactionResult = resolveAssociatedModels(transactionResult)
-
-        guard modelSchema.isSyncable, let syncEngine = self.syncEngine else {
-            if !modelSchema.isSystem {
-                log.error("Unable to sync model (\(modelSchema.name)) where isSyncable is false")
+                                                          deleteInput: .withFilter(filter))
+        switch transactionResult {
+        case .success(let queryAndDeleteResult):
+            guard !queryAndDeleteResult.deletedModels.isEmpty else {
```

Comment:
Should this block be at line 318, before `guard #available...`? Compared to the previous code

Reply:
the previous logic
1. if the model is not syncable, then return the results (will be either success of failure) [L319]
2. if the model is syncable, return if failure [L325]
3. if the model is syncable, and success, check if iOS13, otherwise return the success [L349]
4. if the model is syncable, and success, check if iOS13, then sync the models if there are any. [L340] and return the results [L330]
5. 4. if the model is syncable, and success, check if iOS13, no models to sync, return the success [L338]

The current logic
1. if the result is failure, return it [L340]
2. if the result is success, and there are no models to sync, return empty result [L305]
3. if the result is success, and is not syncable, return results [L316]
4. if the result is success, and is syncable, and is not iOS 13, return [L321]
6. if the result is success, and is syncable, and is iOS 13, then sync the models [L325]

the first difference between the previous and current logic is the current logic checks the transactionResult first before checking whether it should sync or not. Then it checks if it should sync or not based on the three conditions
- are there any models? [L304]
- is the model syncable? [L309]
- is it IOS 13? [L320]

the reason why i pulled this check up [previously L337] is because if we know there are no models to sync, there is no point in checking if sync is required